File: scripts/ai-workflow/phases/test_scenario.py
"""Phase 3: テストシナリオフェーズ

要件定義書と設計書から、Phase 2で決定されたテスト戦略に基づいて
テストシナリオを作成する。
"""
import re
from pathlib import Path
from typing import Dict, Any, List
from .base_phase import BasePhase


class TestScenarioPhase(BasePhase):
    """テストシナリオフェーズ"""

    def __init__(self, *args, **kwargs):
        """初期化"""
        super().__init__(
            phase_name='test_scenario',
            *args,
            **kwargs
        )

    def execute(self) -> Dict[str, Any]:
        """
        テストシナリオフェーズを実行

        Returns:
            Dict[str, Any]: 実行結果
                - success: bool
                - output: str - test-scenario.mdのパス
                - error: Optional[str]
        """
        try:
            # Issue情報を取得
            issue_number = int(self.metadata.data['issue_number'])
            issue_info = self.github.get_issue_info(issue_number)
            issue_info_text = self._format_issue_info(issue_info)

            # 要件定義書と設計書を読み込み
            requirements_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '01_requirements' / 'output' / 'requirements.md'
            design_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '02_design' / 'output' / 'design.md'

            if not requirements_file.exists():
                return {
                    'success': False,
                    'output': None,
                    'error': f'要件定義書が見つかりません: {requirements_file}'
                }

            if not design_file.exists():
                return {
                    'success': False,
                    'output': None,
                    'error': f'設計書が見つかりません: {design_file}'
                }

            # テスト戦略を取得（Phase 2で決定済み）
            test_strategy = self.metadata.data['design_decisions'].get('test_strategy')

            if not test_strategy:
                return {
                    'success': False,
                    'output': None,
                    'error': 'テスト戦略が設計フェーズで決定されていません。Phase 2を先に実行してください。'
                }

            # 実行プロンプトを読み込み
            execute_prompt_template = self.load_prompt('execute')

            # working_dirからの相対パスを使用
            rel_path_requirements = requirements_file.relative_to(self.claude.working_dir)
            rel_path_design = design_file.relative_to(self.claude.working_dir)

            # プロンプトに情報を埋め込み
            execute_prompt = execute_prompt_template.replace(
                '{requirements_document_path}',
                f'@{rel_path_requirements}'
            ).replace(
                '{design_document_path}',
                f'@{rel_path_design}'
            ).replace(
                '{test_strategy}',
                test_strategy
            ).replace(
                '{issue_info}',
                issue_info_text
            ).replace(
                '{issue_number}',
                str(issue_number)
            )

            # Claude Agent SDKでタスクを実行
            messages = self.execute_with_claude(
                prompt=execute_prompt,
                max_turns=40,
                log_prefix='execute'
            )

            # test-scenario.mdのパスを取得
            output_file = self.output_dir / 'test-scenario.md'

            if not output_file.exists():
                return {
                    'success': False,
                    'output': None,
                    'error': f'test-scenario.mdが生成されませんでした: {output_file}'
                }

            # ステータス更新: BasePhase.run()で実行されるため不要
            # self.metadata.update_phase_status('test_scenario', 'completed', str(output_file))
            # self.post_progress('completed', f'テストシナリオが完了しました: {output_file.name}')

            return {
                'success': True,
                'output': str(output_file),
                'error': None
            }

        except Exception as e:
            # ステータス更新: 失敗
            self.metadata.update_phase_status('test_scenario', 'failed')
            # BasePhase.run()で実行されるため不要
            # self.post_progress('failed', f'テストシナリオが失敗しました: {str(e)}')

            return {
                'success': False,
                'output': None,
                'error': str(e)
            }

    def review(self) -> Dict[str, Any]:
        """
        テストシナリオをレビュー

        Returns:
            Dict[str, Any]: レビュー結果
                - result: str - PASS/PASS_WITH_SUGGESTIONS/FAIL
                - feedback: str
                - suggestions: List[str]
        """
        try:
            # test-scenario.mdを読み込み
            test_scenario_file = self.output_dir / 'test-scenario.md'

            if not test_scenario_file.exists():
                return {
                    'result': 'FAIL',
                    'feedback': 'test-scenario.mdが存在しません。',
                    'suggestions': ['execute()を実行してtest-scenario.mdを生成してください。']
                }

            # 要件定義書と設計書のパス
            issue_number = int(self.metadata.data['issue_number'])
            requirements_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '01_requirements' / 'output' / 'requirements.md'
            design_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '02_design' / 'output' / 'design.md'

            # テスト戦略を取得
            test_strategy = self.metadata.data['design_decisions'].get('test_strategy', 'UNKNOWN')

            # レビュープロンプトを読み込み
            review_prompt_template = self.load_prompt('review')

            # working_dirからの相対パスを使用
            rel_path_test_scenario = test_scenario_file.relative_to(self.claude.working_dir)
            rel_path_design = design_file.relative_to(self.claude.working_dir)
            rel_path_requirements = requirements_file.relative_to(self.claude.working_dir)

            # プロンプトに情報を埋め込み
            review_prompt = review_prompt_template.replace(
                '{test_scenario_document_path}',
                f'@{rel_path_test_scenario}'
            ).replace(
                '{design_document_path}',
                f'@{rel_path_design}'
            ).replace(
                '{requirements_document_path}',
                f'@{rel_path_requirements}'
            ).replace(
                '{test_strategy}',
                test_strategy
            )

            # Claude Agent SDKでレビューを実行
            messages = self.execute_with_claude(
                prompt=review_prompt,
                max_turns=30,
                log_prefix='review'
            )

            # レビュー結果をパース
            review_result = self._parse_review_result(messages)

            # レビュー結果をファイルに保存
            review_file = self.review_dir / 'result.md'
            review_file.write_text(review_result['feedback'], encoding='utf-8')
            print(f"[INFO] レビュー結果を保存: {review_file}")

            # GitHub Issueにレビュー結果を投稿: BasePhase.run()で実行されるため不要
            # self.post_review(
            #     result=review_result['result'],
            #     feedback=review_result['feedback'],
            #     suggestions=review_result.get('suggestions')
            # )

            return review_result

        except Exception as e:
            return {
                'result': 'FAIL',
                'feedback': f'レビュー中にエラーが発生しました: {str(e)}',
                'suggestions': []
            }

    def revise(self, review_feedback: str) -> Dict[str, Any]:
        """
        レビュー結果を元にテストシナリオを修正

        Args:
            review_feedback: レビュー結果のフィードバック

        Returns:
            Dict[str, Any]: 修正結果
                - success: bool
                - output: str - test-scenario.mdのパス
                - error: Optional[str]
        """
        try:
            # 元のテストシナリオを読み込み
            test_scenario_file = self.output_dir / 'test-scenario.md'

            if not test_scenario_file.exists():
                return {
                    'success': False,
                    'output': None,
                    'error': 'test-scenario.mdが存在しません。'
                }

            # 要件定義書と設計書のパス
            issue_number = int(self.metadata.data['issue_number'])
            requirements_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '01_requirements' / 'output' / 'requirements.md'
            design_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '02_design' / 'output' / 'design.md'

            # テスト戦略を取得
            test_strategy = self.metadata.data['design_decisions'].get('test_strategy', 'UNKNOWN')

            # 修正プロンプトを読み込み
            revise_prompt_template = self.load_prompt('revise')

            # working_dirからの相対パスを使用
            rel_path_test_scenario = test_scenario_file.relative_to(self.claude.working_dir)
            rel_path_design = design_file.relative_to(self.claude.working_dir)
            rel_path_requirements = requirements_file.relative_to(self.claude.working_dir)

            # プロンプトに情報を埋め込み
            revise_prompt = revise_prompt_template.replace(
                '{test_scenario_document_path}',
                f'@{rel_path_test_scenario}'
            ).replace(
                '{review_feedback}',
                review_feedback
            ).replace(
                '{design_document_path}',
                f'@{rel_path_design}'
            ).replace(
                '{requirements_document_path}',
                f'@{rel_path_requirements}'
            ).replace(
                '{test_strategy}',
                test_strategy
            ).replace(
                '{issue_number}',
                str(issue_number)
            )

            # Claude Agent SDKでタスクを実行
            messages = self.execute_with_claude(
                prompt=revise_prompt,
                max_turns=40,
                log_prefix='revise'
            )

            # test-scenario.mdのパスを取得
            output_file = self.output_dir / 'test-scenario.md'

            if not output_file.exists():
                return {
                    'success': False,
                    'output': None,
                    'error': '修正されたtest-scenario.mdが生成されませんでした。'
                }

            return {
                'success': True,
                'output': str(output_file),
                'error': None
            }

        except Exception as e:
            return {
                'success': False,
                'output': None,
                'error': str(e)
            }

    def _format_issue_info(self, issue_info: Dict[str, Any]) -> str:
        """
        Issue情報をフォーマット

        Args:
            issue_info: Issue情報

        Returns:
            str: フォーマットされたIssue情報
        """
        formatted = f"""
## Issue情報

- **Issue番号**: #{issue_info['number']}
- **タイトル**: {issue_info['title']}
- **状態**: {issue_info['state']}
- **URL**: {issue_info['url']}
- **ラベル**: {', '.join(issue_info['labels']) if issue_info['labels'] else 'なし'}

### 本文

{issue_info['body']}
"""
        return formatted.strip()

File: scripts/ai-workflow-v2/src/phases/test-scenario.ts
import fs from 'fs-extra';
import path from 'node:path';
import { BasePhase, type PhaseInitializationParams } from './base-phase.js';
import { PhaseExecutionResult } from '../types.js';

type IssueInfo = {
  number: number;
  title: string;
  state: string;
  url: string;
  labels: string[];
  body: string;
};

export class TestScenarioPhase extends BasePhase {
  constructor(params: PhaseInitializationParams) {
    super({ ...params, phaseName: 'test_scenario' });
  }

  protected async execute(): Promise<PhaseExecutionResult> {
    const issueInfo = (await this.getIssueInfo()) as IssueInfo;
    const planningReference = this.getPlanningDocumentReference(issueInfo.number);

    const requirementsFile = this.getPhaseOutputFile('requirements', 'requirements.md', issueInfo.number);
    const designFile = this.getPhaseOutputFile('design', 'design.md', issueInfo.number);

    // requirements と design はオプショナル（Issue #405）
    let requirementsReference: string;
    if (requirementsFile) {
      const ref = this.getAgentFileReference(requirementsFile);
      requirementsReference = ref ?? '要件定義書は利用できません。Planning情報とIssue情報から要件を推測してください。';
    } else {
      requirementsReference = '要件定義書は利用できません。Planning情報とIssue情報から要件を推測してください。';
    }

    let designReference: string;
    if (designFile) {
      const ref = this.getAgentFileReference(designFile);
      designReference = ref ?? '設計ドキュメントは利用できません。Planning情報から設計を推測してください。';
    } else {
      designReference = '設計ドキュメントは利用できません。Planning情報から設計を推測してください。';
    }

    // test_strategy もオプショナル（Issue #405）
    const testStrategy = this.metadata.data.design_decisions.test_strategy ??
      'テスト戦略は設定されていません。要件と設計から適切なテスト戦略を決定してください。';

    const executePrompt = this.loadPrompt('execute')
      .replace('{planning_document_path}', planningReference)
      .replace('{requirements_document_path}', requirementsReference)
      .replace('{design_document_path}', designReference)
      .replace('{test_strategy}', testStrategy)
      .replace('{issue_info}', this.formatIssueInfo(issueInfo))
      .replace('{issue_number}', String(issueInfo.number));

    await this.executeWithAgent(executePrompt, { maxTurns: 40 });

    const scenarioFile = path.join(this.outputDir, 'test-scenario.md');
    if (!fs.existsSync(scenarioFile)) {
      return {
        success: false,
        error: `test-scenario.md が見つかりません: ${scenarioFile}`,
      };
    }

    // Phase outputはPRに含まれるため、Issue投稿は不要（Review resultのみ投稿）
    // try {
    //   const content = fs.readFileSync(scenarioFile, 'utf-8');
    //   await this.postOutput(content, 'テストシナリオ');
    // } catch (error) {
    //   const message = (error as Error).message ?? String(error);
    //   console.warn(`[WARNING] GitHub へのテストシナリオ投稿に失敗しました: ${message}`);
    // }

    return {
      success: true,
      output: scenarioFile,
    };
  }

  protected async review(): Promise<PhaseExecutionResult> {
    const issueInfo = (await this.getIssueInfo()) as IssueInfo;
    const scenarioFile = path.join(this.outputDir, 'test-scenario.md');

    if (!fs.existsSync(scenarioFile)) {
      return {
        success: false,
        error: 'test-scenario.md が存在しません。execute() を先に実行してください。',
      };
    }

    const requirementsFile = this.getPhaseOutputFile('requirements', 'requirements.md', issueInfo.number);
    const designFile = this.getPhaseOutputFile('design', 'design.md', issueInfo.number);

    const scenarioReference = this.getAgentFileReference(scenarioFile);
    if (!scenarioReference) {
      return {
        success: false,
        error: 'Agent が test-scenario.md を参照できません。',
      };
    }

    // requirements と design はオプショナル（Issue #405）
    let requirementsReference: string;
    if (requirementsFile) {
      const ref = this.getAgentFileReference(requirementsFile);
      requirementsReference = ref ?? '要件定義書は利用できません。テストシナリオから要件を推測してレビューしてください。';
    } else {
      requirementsReference = '要件定義書は利用できません。テストシナリオから要件を推測してレビューしてください。';
    }

    let designReference: string;
    if (designFile) {
      const ref = this.getAgentFileReference(designFile);
      designReference = ref ?? '設計ドキュメントは利用できません。テストシナリオから設計を推測してレビューしてください。';
    } else {
      designReference = '設計ドキュメントは利用できません。テストシナリオから設計を推測してレビューしてください。';
    }

    const testStrategy = this.metadata.data.design_decisions.test_strategy ??
      'テスト戦略は設定されていません。テストシナリオ内容から適切なテスト観点でレビューしてください。';

    const reviewPrompt = this.loadPrompt('review')
      .replace('{test_scenario_document_path}', scenarioReference)
      .replace('{design_document_path}', designReference)
      .replace('{requirements_document_path}', requirementsReference)
      .replace('{test_strategy}', testStrategy);

    const messages = await this.executeWithAgent(reviewPrompt, { maxTurns: 30, logDir: this.reviewDir });
    const reviewResult = await this.contentParser.parseReviewResult(messages);

    const reviewFile = path.join(this.reviewDir, 'result.md');
    fs.writeFileSync(reviewFile, reviewResult.feedback, 'utf-8');

    await this.github.postReviewResult(
      issueInfo.number,
      this.phaseName,
      reviewResult.result,
      reviewResult.feedback,
      reviewResult.suggestions,
    );

    return {
      success: reviewResult.result !== 'FAIL',
      output: reviewResult.result,
      error: reviewResult.result === 'FAIL' ? reviewResult.feedback : undefined,
    };
  }

  public async revise(reviewFeedback: string): Promise<PhaseExecutionResult> {
    const issueInfo = (await this.getIssueInfo()) as IssueInfo;
    const scenarioFile = path.join(this.outputDir, 'test-scenario.md');

    if (!fs.existsSync(scenarioFile)) {
      return {
        success: false,
        error: 'test-scenario.md が存在しません。execute() を先に実行してください。',
      };
    }

    const requirementsFile = this.getPhaseOutputFile('requirements', 'requirements.md', issueInfo.number);
    const designFile = this.getPhaseOutputFile('design', 'design.md', issueInfo.number);

    const scenarioReference = this.getAgentFileReference(scenarioFile);
    if (!scenarioReference) {
      return {
        success: false,
        error: 'Agent が test-scenario.md を参照できません。',
      };
    }

    // requirements と design はオプショナル（Issue #405）
    let requirementsReference: string;
    if (requirementsFile) {
      const ref = this.getAgentFileReference(requirementsFile);
      requirementsReference = ref ?? '要件定義書は利用できません。テストシナリオから要件を推測してください。';
    } else {
      requirementsReference = '要件定義書は利用できません。テストシナリオから要件を推測してください。';
    }

    let designReference: string;
    if (designFile) {
      const ref = this.getAgentFileReference(designFile);
      designReference = ref ?? '設計ドキュメントは利用できません。テストシナリオから設計を推測してください。';
    } else {
      designReference = '設計ドキュメントは利用できません。テストシナリオから設計を推測してください。';
    }

    const testStrategy = this.metadata.data.design_decisions.test_strategy ??
      'テスト戦略は設定されていません。テストシナリオ内容から適切なテスト戦略を決定してください。';

    const revisePrompt = this.loadPrompt('revise')
      .replace('{test_scenario_document_path}', scenarioReference)
      .replace('{design_document_path}', designReference)
      .replace('{requirements_document_path}', requirementsReference)
      .replace('{test_strategy}', testStrategy)
      .replace('{review_feedback}', reviewFeedback)
      .replace('{issue_number}', String(issueInfo.number));

    await this.executeWithAgent(revisePrompt, { maxTurns: 40, logDir: this.reviseDir });

    if (!fs.existsSync(scenarioFile)) {
      return {
        success: false,
        error: '改訂後の test-scenario.md を確認できませんでした。',
      };
    }

    return {
      success: true,
      output: scenarioFile,
    };
  }
}
